gobblin-metrics/src/main/java/gobblin/metrics/ContextAwareCounter.java

```diff
@@ -0,0 +1,132 @@
+/* (c) 2014 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.metrics;
+
+import java.util.Collection;
+import java.util.List;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.MetricRegistry;
+
+import com.google.common.base.Optional;
+
+
+/**
+ * A type of {@link Counter}s that are aware of their {@link MetricContext} and can have associated
+ * {@link Tag}s.
+ *
+ * <p>
+ *   Any updates to a {@link ContextAwareCounter} will be applied automatically to the
+ *   {@link ContextAwareCounter} of the same name in the parent {@link MetricContext}.
+ * </p>
+ *
+ * <p>
+ *   This class wraps a {@link com.codahale.metrics.Counter} and delegates calls to public methods of
+ *   {@link com.codahale.metrics.Counter} to the wrapped {@link com.codahale.metrics.Counter}.
+ * </p>
+ *
+ * @author ynli
+ */
+class ContextAwareCounter extends Counter implements ContextAwareMetric {
+
+  private final String name;
+  private final MetricContext context;
+  private final Counter counter;
```

Comment:
Since everything is public in the Metrics implementations, can we just extend counter and override some methods instead of wrapping a Counter instance? That way a lot of methods will not need to be reimplemented (e.g. getCount).